### ISSUE TITLE
synthetic/sum_is_commutative_klee_float_bug.x86_64 is missing counter examples

### ISSUE THREAD

**Issue by delcypher** [delcypher/fp-benchmarks-imperial#4]
```
WARNING:MainThread: MATCH SPEC for task no_invalid_deref but with warnings:
/home/user/build_O2/benchmarks/c/imperial/synthetic/sum_is_commutative_klee_float_bug.x86_64.bc (/home/dsl11/dev/klee-afr/experiments/first/wd/workdir-83/klee-wd)
Observed counter examples not listed in spec
  Test 000006 in /home/user/fp-bench/benchmarks/c/imperial/synthetic/sum_is_commutative/sum_is_commutative.c:40
```

@afd This is your benchmark. Could you added the out of bound memory access to the list of counter examples for `no_invalid_deref` verification task.

**Comment by afd** [comment]
Hi Dan

There is not supposed to be an out of bounds access for this benchmark, and looking at the code I cannot see how one can occur.  Are you able to point to where the benchmark is wrong?  If you can, the solution should be to fix the program, not change its spec.

**Comment by delcypher** [comment]
@afd The `spec.yml` file lists `no_invalid_deref` as expected to be incorrect.

```
  klee_float_bug:
    categories:
      - issta_2017
    defines:
      FLOAT_TYPE: float
      KLEE: null
      N: "5"
      BUG: "1"
    dependencies:
      klee_runtime: {}
    verification_tasks:
      no_assert_fail:
        correct: false
        counter_examples:
          -
            description: >
              An injected bug messes up one of the loop conditions,
              which will stop the sums from matching
            locations:
              -
                file: "sum_is_commutative.c" 
                line: 72

      no_invalid_deref:
        correct: false
        description: >
          The messed up bounds may cause and out-of-bounds access
```

**Comment by delcypher** [comment]
The bug is in the loop body

```c
FLOAT_TYPE sum2(FLOAT_TYPE data[N]) {
  FLOAT_TYPE sum = (FLOAT_TYPE)0.0f;
  for(int i = 0; i <
#if BUG==0
 N
#else
 data[i]
#endif
       ; i++) {
    sum = data[i] + sum;
  }
  return sum;
}
```

If symbolic data is used as the loop bound then eventually `i` will be incremented enough to access `data` out of bounds.

**Comment by afd** [comment]
Ah, got it.  For some reason the file I looked at in the repo didn't 
have this - must have been looking at the wrong file or something. I can 
deal with this on Friday, or please feel free to add the clause yourself 
in the meantime.  (I don't want to do it in a rush now in case I get it 
wrong.)

Ally

**Comment by delcypher** [comment]
Hmm I was going to apply the change but that line number is wrong. The access is on line 44 not 40. This might be a bug in Clang. Perhaps during optimization the correct debug information gets lost.

**Comment by delcypher** [comment]
Hmm with Clang 3.4 the debug information is wrong at `-O0`. Checking newer clang now.

**Comment by delcypher** [comment]
In Clang 3.9 the location is correct so there's no need to report the issue upstream. I'll just rework the benchmark.

**Comment by delcypher** [comment]
This has been fixed by d0347079540e250ef71fbbd70c4dcfd6d0f6cce9
